File: backend/src/main/java/com/SmartCommerce/service/UserService.java
package com.SmartCommerce.service;

import com.SmartCommerce.dto.*;

import java.util.List;

public interface UserService {

    UserDTO register(RegisterRequestDTO request);

    UserDTO createUserByAdmin(CreateUserByAdminRequestDTO request);

    UserDTO getUserById(Long id);

    UserDTO getUserByEmail(String email);

    List<UserDTO> getAllUsers();

    List<UserDTO> getAllUsersIncludingInactive();

    List<UserDTO> getActiveUsers();

    UserDTO updateUser(Long id, UpdateUserRequestDTO request);

    UserDTO adminUpdateUser(Long id, AdminUpdateUserRequestDTO request);

    void changePassword(Long id, ChangePasswordRequestDTO request);

    void deactivateUser(Long id);

    void activateUser(Long id);

    void deleteUser(Long id);

    List<UserDTO> searchUsersByName(String name);

    List<UserDTO> getUsersByRole(String role);
}

File: frontend/src/services/userService.js
import api from './api.js';
import { createUserDTO, createUpdateUserRequest, createChangePasswordRequest } from '../types/user.js';

export const userApi = {
    getAllUsers: async () => {
        const response = await api.get('/users');
        return response.data.map(user => createUserDTO(user));
    },

    getActiveUsers: async () => {
        const response = await api.get('/users/active');
        return response.data.map(user => createUserDTO(user));
    },

    getUserById: async (id) => {
        const response = await api.get(`/users/${id}`);
        return createUserDTO(response.data);
    },

    getUserByEmail: async (email) => {
        const response = await api.get(`/users/email/${encodeURIComponent(email)}`);
        return createUserDTO(response.data);
    },

    createUser: async (userData) => {
        const response = await api.post('/users', userData);
        return createUserDTO(response.data);
    },

    updateUser: async (id, userData) => {
        const request = createUpdateUserRequest(userData);
        const response = await api.put(`/users/${id}`, request);
        return createUserDTO(response.data);
    },

    adminUpdateUser: async (id, userData) => {
        const response = await api.put(`/users/${id}`, userData);
        return createUserDTO(response.data);
    },

    changePassword: async (id, passwordData) => {
        const request = createChangePasswordRequest(passwordData);
        await api.put(`/users/${id}/password`, request);
    },

    deleteUser: async (id) => {
        await api.delete(`/users/${id}`);
    },

    deactivateUser: async (id) => {
        await api.delete(`/users/${id}`);
    },

    activateUser: async (id) => {
        await api.post(`/users/${id}/activate`);
    },

    searchUsers: async (name) => {
        const response = await api.get('/users/search', { params: { name } });
        return response.data.map(user => createUserDTO(user));
    },

    getUsersByRole: async (role) => {
        const response = await api.get(`/users/role/${role}`);
        return response.data.map(user => createUserDTO(user));
    }
};

export default userApi;
